README.md
# 열린옷장 자원봉사 #

[![Build Status](https://travis-ci.org/opencloset/volunteer.svg?branch=v0.4.15)](https://travis-ci.org/opencloset/volunteer)

열린옷장의 자원봉사에 관한 서비스입니다.

# 설치 및 실행 #

    $ cpanm --installdeps .
    $ bower install
    $ cp volunteer.conf.sample volunteer.conf
    $ MOJO_CONFIG=volunteer.conf morbo -l 'http://*:5000' ./script/volunteer    # http://localhost:5000


## docs ##

    $ npm install -g grunt-cli
    $ npm install
    $ gem install bundler
    $ bundle install
    $ bundle exec jekyll serve    # http://localhost:5000

# 환경변수 #

- OPENCLOSET_APISTORE_ID
- OPENCLOSET_APISTORE_API_STORE_KEY
- OPENCLOSET_GOOGLE_PRIVATE_KEY

google private key 는 암호화된 JSON Web Token 파일입니다.
자세한 내용은 [Service Account](https://developers.google.com/identity/protocols/OAuth2ServiceAccount)를 참고.

## Build docker images ##

    $ docker build -f Dockerfile -t opencloset/volunteer .
    $ docker build -f Dockerfile.minion -t opencloset/volunteer/minion .


Changes
# Change Log
The format is based on [Keep a Changelog](http://keepachangelog.com/)

## [Unreleased]
### Added
- 자원봉사 신청을 제한함 (#124)

## [v0.4.15] - Mon, 30 Nov 2020 09:55:36 +0900
### Changed
- 같은 날 여러번 신청 가능하게 함 (#120)

## [v0.4.14] - Sun, 29 Nov 2020 10:02:17 +0900
### Fixed
- 승인 이후에 google calendar 에 스케쥴 등록 (#120)

## [v0.4.13] - Sat, 28 Nov 2020 11:35:54 +0900
### Changed
- 신청 간소화 (#120)
  - staff 이 대행
  - 신청 -> 승인 절차 대신에 바로 승인
- 활동시간 변경 (#120)

## [v0.4.12] - Thu, 26 Mar 2020 20:25:12 +0900
### Fixed
- ~12, 12~ 와 같은 중첩된 시간대를 구분할 수 있도록 함 (#116)

## [v0.4.11] - Thu, 27 Feb 2020 21:31:44 +0900
### Changed
- 봉사활동 시간대에서 09-16 을 제거 (#114)

## [v0.4.10] - Mon, 11 Nov 2019 23:11:37 +0900
### Fixed
- 봉사활동 시간대 오류 수정 (#113)

## [v0.4.9] - Thu, 08 Aug 2019 19:12:57 +0900
### Fixed
- 최대봉사인원이 0 일때를 고려함 (#113)

## [v0.4.8] - Wed, 17 Jul 2019 20:58:55 +0900
### Changed
- 봉사활동 시간대 변경 (#113)

## [v0.4.7] - Tue, 07 May 2019 21:06:33 +0900

### Changed
- 개인정보취급방침 변경 (#111)

## [v0.4.6] - Mon, 11 Jun 2018 13:01:46 +0900

### Changed
- Google API 인증 실패시 3번까지 재시도 (#109)

## [v0.4.5] - Fri, 04 May 2018 21:02:48 +0900

### Fixed
- Set default config file in docker image

## [v0.4.4] - Fri, 04 May 2018 16:00:44 +0900

### Changed
- docker image 에서 private key file 을 제거(대신에 bind mount 를 사용)

## [v0.4.3] - Fri, 04 May 2018 15:36:45 +0900

### Added
- travis-ci 에서 docker image 를 build

## [v0.4.2] - Tue, 23 Jan 2018 12:24:07 +0900

### Changed
- 열린옷장 위치안내 SMS 수정 (#72)

## [v0.4.1] - Tue, 14 Nov 2017 17:14:51 +0900

### Fixed
- 일별 봉사신청가능시간 버그 수정 (#106)

## [v0.4.0] - Thu, 02 Nov 2017 16:52:46 +0900

### Added
- dockerize

## [v0.3.13] - Tue, 22 Aug 2017 17:16:09 +0900

### Changed
- 첫 소개화면 이미지 변경 (#103)

## [v0.3.12] - Wed, 16 Aug 2017 16:04:07 +0900

### Added
- calendar_insert helper

### Chaned
- quickAdd 대신에 calendar_insert 를 사용해서 캘린더 이벤트를 등록

## [v0.3.11] - Wed, 16 Aug 2017 15:12:44 +0900

### Fixed
- Fix typo at google calendar quickadd text

## [v0.3.10] - Wed, 16 Aug 2017 15:05:49 +0900

### Changed
- google calendar quickadd text format

## [v0.3.9] - Wed, 16 Aug 2017 12:11:14 +0900

### Changed
- 봉사활동 시간변경 (#102)
  - 활동 시간을 시간 단위가 아닌 분단위까지 고려
  - 문자 메세지 및 구글캘린더에 분단위 까지 표시

## [v0.3.8] - Wed, 09 Aug 2017 15:49:00 +0900

### Changed
- 봉사활동 시간변경 (#100)
  - 10시 ~ 13시 3시간 -> 10시 ~ 12시 30분
  - 10시 ~ 18시 7시간 -> 10시 ~ 18시 6시간 30분

## [v0.3.7] - Mon, 07 Aug 2017 13:30:28 +0900

### Changed
- 첫화면에서 열린사진관 관련된 내용을 삭제 (#97)

### Fixed
- 승인/위치안내 문자의 수신번호 포맷문제로 인한 오류 수정 (#101)

## [v0.3.6] - Mon, 19 Jun 2017 13:52:54 +0900
### Changed
- 자원봉사자의 사진 미리보기를 더 크게 함 (#15)

## [v0.3.5] - Sat, 17 Jun 2017 15:56:26 +0900
### Added
- 자원봉사자의 사진을 업로드 할 수 있도록 함 (#15)

## [v0.3.4] - Thu, 08 Jun 2017 04:05:38 +0900
### Added
- 자원봉사자에 대한 코멘트를 입력할 수 있는 기능 추가

### Changed
- sms 를 직접 보내는 대신 `sms` helper 를 활용

## [v0.3.3] - Thu, 08 Jun 2017 02:16:43 +0900
### Added
- staff 의 봉사활동 목록을 /works 로 이관
- Add helper `phone_format`
- private_routes 추가
  - GET /works

### Removed
- Access_control_CORS 삭제
  - OPTIONS /status route 삭제
  - OPTIONS /1365 route 삭제

## [v0.3.2] - Thu, 01 Jun 2017 16:35:29 +0900
### Changed
- staff 계정은 기간제한 없이 봉사활동을 신청할 수 있도록 함

v0.3.1    Wed, 19 Apr 2017 17:55:52 +0900
  - 신청서에 복수신청에 관한 안내 추가 (#51)
  - 18:00 이후에는 익일 신청을 할 수 없도록 함 (#88)
  - intro 의 좌우 swap (#82)

v0.3.0    Thu, 13 Apr 2017 11:53:26 +0900
  - 한번에 여러날을 신청할 수 있도록 함 (#51)

v0.2.20   Wed, 22 Mar 2017 12:32:29 +0900
  - 요일/시간대별로 인원수를 제한할 수 있는 기능 추가 (#85)

v0.2.19   Wed, 22 Mar 2017 10:58:39 +0900
  - 소개화면 내용 변경 (#82)

v0.2.18   Thu, 09 Mar 2017 12:30:48 +0900
  - 소개화면 내용 변경 (#82)

v0.2.17   Wed, 08 Mar 2017 14:01:52 +0900
  - footer 를 추가
  - 소개화면 내용 변경 (#82)

v0.2.16   Tue, 07 Mar 2017 12:02:31 +0900
  - 봉사활동 소개 화면을 변경 (#82)

v0.2.15   Mon, 13 Feb 2017 11:22:44 +0900
  - 1365 소속센터를 입력받지 않음 (#80)

v0.2.14   Fri, 20 Jan 2017 15:42:53 +0900
  - 대문사진변경 (#70)

v0.2.13   Mon, 09 Jan 2017 17:36:52 +0900
  - sms 발신번호 변경
  - extra-holidays.ini 의존성을 제거하고 환경변수로 관리

v0.2.12   Wed, 21 Dec 2016 16:55:21 +0900
  - sms 에 포함된 전화번호 변경 (#79)

v0.2.11   Fri, 09 Dec 2016 14:52:00 +0900
  - 다음해의 휴일목록까지 고려함 (#78)

v0.2.10   Tue, 29 Nov 2016 15:46:00 +0900
  - 구글 캘린더 이벤트 등록 오류 수정 (#77)

v0.2.9    Fri, 25 Nov 2016 12:51:29 +0900
  - intro 이미지 변경 (#70)
  - 1365 소속센터 입력란 추가 (#76)
  - 소개이미지의 자동 크기조정

v0.2.8    Thu, 17 Nov 2016 14:40:41 +0900
  - 자원봉사자에게 보내는 문자 메시지에 변경된 사무실 위치를 반영 (#72)

v0.2.7    Fri, 04 Nov 2016 13:43:55 +0900
  - intro 이미지 변경 (#70)

v0.2.6    Fri, 28 Oct 2016 17:17:10 +0900
  - 1365와 관련된 SMS 및 내용을 삭제 (#68)
  - 봉사활동 신청 가능 기간을 +1d ~ +1m 으로 제한 (#65)
  - 17세이상 고등학생부터 봉사신청이 가능하도록 변경 (#66)

v0.2.5    Wed, 19 Oct 2016 15:33:16 +0900
  - 신청일을 +1d ~ +3m 으로 제한 (#65)

v0.2.4    Wed, 28 Sep 2016 10:36:36 +0900
  - 1365 안내 문구 추가

v0.2.3    Mon, 26 Sep 2016 17:53:20 +0900
  - Fix 일요일도 봉사신청을 할 수 있도록 함 (#61)

v0.2.2    Thu, 22 Sep 2016 13:32:12 +0900
  - 일요일도 봉사신청을 할 수 있도록 함 (#61)

v0.2.1    Fri, 09 Sep 2016 11:44:08 +0900
  - 추석연휴일(09-17)에 봉사활동을 신청할 수 없도록 함

v0.2.0    Wed, 15 Jun 2016 12:17:05 +0900
  - 봉사활동기간의 로그인 아이디를 입력받음 (#59)
  - 정기봉사활동에 대해 설명 제거

v0.1.21    Thu, 28 Apr 2016 16:35:08 +0900
  - 같은날 중복신청에 관한 버그 수정 (#58)
  - 봉사활동 신청시에 개인정보수집동의 (#57)

v0.1.20    Thu, 14 Apr 2016 18:49:44 +0900
  - 방문안함 일때 캘린더에서 삭제 (#56)

v0.1.19    Mon, 11 Apr 2016 18:09:44 +0900
  - 같은날 두번이상 신청할 수 없게 함 (#47)

v0.1.18    Fri, 25 Mar 2016 18:28:20 +0900
  - 열린옷장 위치 안내 SMS 수정 (#48)

v0.1.17    Tue, 16 Feb 2016 12:24:02 +0900
  - use opencloset.css
  - 정기봉사활동에 대해 설명

v0.1.16    Wed, 27 Jan 2016 15:01:52 +0900
  - 정기봉사활동에는 신청할 수 없도록 함(매주 화/목 02-11 ~ 06-09)

v0.1.15    Thu, 21 Jan 2016 17:29:01 +0900
  - Add home with intro

v0.1.14    Wed, 20 Jan 2016 14:50:25 +0900
  - 제헌절은 공휴일이 아니라서 예약가능한 날로 변경
  - 동시간대 봉사자 수를 설정파일에서 조정

v0.1.13    Wed, 30 Dec 2015 17:14:26 +0900
  - calc able hours skip canceled status
  - use gruntfile rather than mojo asset pack
  - 2016년 휴일을 예약가능하지 않도록 수정
  - 활동 승인후에 네이버 밴드 안내 문자 메세지 제거

v0.1.12    Wed, 14 Oct 2015 18:22:46 +0900
  - 수정할때에 더 다양한 봉사활동 시간을 제공


lib/OpenCloset/Volunteer/Controller/Work.pm
package OpenCloset::Volunteer::Controller::Work;
use Mojo::Base 'Mojolicious::Controller';

use Data::Pageset;
use DateTime::Format::ISO8601;
use DateTime;
use Email::Simple;
use Encode qw/encode_utf8/;
use HTML::FillInForm::Lite;
use String::Random ();

has schema => sub { shift->app->schema };

=head1 METHODS

=head2 add

    GET /works/new

=cut

sub add {
    my $self = shift;

    my $now = DateTime->now( time_zone => $self->config->{timezone} );
    my $year = $now->year;

    my $user_id = $self->session('access_token');
    my $staff;
    if ($user_id) {
        my $user = $self->schema->resultset('User')->find( { id => $user_id } );
        my $user_info = $user->user_info;
        $staff = $user_info->staff;
    }

    $self->render( now => $now, holidays => [$self->holidays( $year, $year + 1 )], staff => $staff );
}

=head2 create

    POST /works

=cut

sub create {
    my $self = shift;
    my $v    = $self->validation;
    $self->_validate_volunteer($v);
    $self->_validate_volunteer_work($v);
    if ( $v->has_error ) {
        my $failed = $v->failed;
        return $self->error( 400, 'Parameter Validation Failed: ' . join( ', ', @$failed ) );
    }

    my $name = $v->param('name');
    my $phone = $v->param('phone');
    my $activity_datetime = $v->every_param('activity-datetime');

    my @added;
    my $tz = $self->config->{timezone};
    my $volunteer;
    for my $datetime (@$activity_datetime) {
        my ( $date, $hours ) = split / /, $datetime;
        my ( $from, $to )    = split /-/, $hours;

        my $able_hours = $self->_able_hour($date);
        return $self->error( 400, "Not allow activity hours: $hours" ) unless $able_hours->{$hours};

        my $dt = DateTime::Format::ISO8601->parse_datetime($date);
        $dt->set_time_zone($tz);

        $volunteer = $self->schema->resultset('Volunteer')->find_or_create(
            {
                name  => $name,
                phone => $phone,
            }
        );

        return $self->error( 500, 'Failed to find or create Volunteer' ) unless $volunteer;

        my $work = $self->schema->resultset('VolunteerWork')->create(
            {
                volunteer_id       => $volunteer->id,
                activity_from_date => "$date $from:00",
                activity_to_date   => "$date $to:00",
                status             => 'approved',
                authcode           => String::Random->new->randregex('[a-zA-Z0-9]{32}')
            }
        );

        return $self->error( 500, 'Failed to create Volunteer Work' ) unless $work;

        ## Google Calendar
        my $from_date = $work->activity_from_date;
        my $to_date   = $work->activity_to_date;
        my $event_id  = $self->calendar_insert( $volunteer->name, $from_date, $to_date );
        $work->update( { event_id => $event_id } );
        push @added, sprintf( '%d월 %d일', $dt->month, $dt->day );
    }

    my $msg = $self->render_to_string( 'sms/status-reported', format => 'txt', name => $name, dates => \@added );
    $phone =~ s/-//g;
    $self->sms( $phone, $msg );

    $self->render( 'work/done', volunteer => $volunteer, dates => \@added );
}

=head2 find_work

    under /works/:id

=cut

sub find_work {
    my $self = shift;
    my $id   = $self->param('id');

    my $work = $self->schema->resultset('VolunteerWork')->find($id);
    unless ($work) {
        $self->error( 404, "Not found volunteer work: $id" );
        return;
    }

    $self->stash( work => $work );
    return 1;
}

=head2 work

    # work
    GET /works/:id

=cut

sub work {
    my $self = shift;
    my $work = $self->stash('work');

    my $volunteer = $work->volunteer;
    my $works     = $volunteer->volunteer_works( { id => { '!=' => $work->id } } );
    my $guestbook = $work->volunteer_guestbooks->next;

    $self->render( works => [$works->all], guestbook => $guestbook );
}

=head2 cancel

    # work.cancel
    GET /works/:id/cancel?phone=xxxx (뒷자리)

=cut

sub cancel {
    my $self      = shift;
    my $work      = $self->stash('work');
    my $phone     = $self->param('phone');
    my $volunteer = $work->volunteer;

    return $self->error( 400, '본인확인을 할 수 없습니다.' ) unless $phone;
    return $self->error( 400, '신청자의 휴대폰번호와 일치하지 않습니다.' )
        if substr( $volunteer->phone, -4 ) ne $phone;

    $self->stash( volunteer => $volunteer );
}

=head2 edit

    # work.edit
    GET /works/:id/edit

=cut

sub edit {
    my $self      = shift;
    my $authcode  = $self->param('authcode') || '';
    my $work      = $self->stash('work');
    my $volunteer = $work->volunteer;

    return $self->error( 400, 'Wrong authcode' ) if $authcode ne $work->authcode;

    my $now  = DateTime->now;
    my $year = $now->year;
    my $from = $work->activity_from_date;
    my $to   = $work->activity_to_date;

    my %filled = ( $work->get_columns, $volunteer->get_columns );
    $filled{reason} = [split /\|/, $filled{reason}];
    $filled{path}   = [split /\|/, $filled{path}];
    $filled{'activity-date'}  = $from->ymd;
    $filled{'activity-hours'} = $from->hour . '-' . $to->hour;
    $filled{birth_date} = $volunteer->birth_date->ymd if $volunteer->birth_date;
    $self->stash( holidays => [$self->holidays( $year, $year + 1 )] );

    my $html = $self->render_to_string( 'work/edit', format => 'html' );
    my $fill = HTML::FillInForm::Lite->new;
    $self->render( text => $fill->fill( \$html, \%filled ), format => 'html' );
}

=head2 update

    # work.update
    POST /works/:id

=cut

sub update {
    my $self      = shift;
    my $authcode  = $self->param('authcode') || '';
    my $work      = $self->stash('work');
    my $volunteer = $work->volunteer;

    return $self->error( 400, error => 'Wrong authcode' ) if $authcode ne $work->authcode;

    my $v = $self->validation;
    $self->_validate_volunteer_work($v);
    if ( $v->has_error ) {
        my $failed = $v->failed;
        return $self->error( 400, 'Parameter Validation Failed: ' . join( ', ', @$failed ) );
    }

    my $activity_date  = $v->param('activity-date');
    my $activity_hours = $v->param('activity-hours');
    my $need_1365      = $v->param('need_1365');
    my $org_username   = $v->param('org_username');
    my $period         = $v->param('period');
    my $talent         = $v->param('talent');
    my $comment        = $v->param('comment');
    my $activity       = $v->param('activity');
    my $reasons        = $v->every_param('reason');
    my $paths          = $v->every_param('path');
    my $job            = $v->param('job');
    my ( $from, $to ) = split /-/, $activity_hours;

    $work->update(
        {
            activity_from_date => "$activity_date $from:00",
            activity_to_date   => "$activity_date $to:00",
            need_1365          => $need_1365,
            org_username       => $org_username,
            period             => $period,
            reason             => join( '|', @$reasons ),
            path               => join( '|', @$paths ),
            job                => $job,
            activity           => $activity,
            talent             => $talent,
            comment            => $comment,
        }
    );

    if ( $work->event_id ) {
        $self->delete_event( $work->event_id );
        my $from     = $work->activity_from_date;
        my $to       = $work->activity_to_date;
        my $event_id = $self->calendar_insert( $volunteer->name, $from, $to );
        $work->update( { event_id => $event_id } );
    }

    $self->render( 'work/done', work => $work );
}

=head2 update_status

    PUT /works/:id/status?status=reported|approved|done|canceled

=cut

sub update_status {
    my $self      = shift;
    my $work      = $self->stash('work');
    my $volunteer = $work->volunteer;

    my $validation = $self->validation;
    $validation->required('status')->in(qw/reported approved done canceled drop/);
    return $self->error( 400, 'Parameter Validation Failed' ) if $validation->has_error;

    my $from_status = $work->status;
    my $status      = $validation->param('status');
    $work->update( { status => $status } );

    if ( $status eq 'approved' ) {
        my $phone = $volunteer->phone =~ s/\-//gr;
        my $msg = $self->render_to_string( 'sms/status-approved', format => 'txt', work => $work );
        chomp $msg;
        my $sent = $self->sms( $phone, $msg );
        $self->log->error("Failed to send SMS: $msg") unless $sent;

        $msg = $self->render_to_string( 'sms/opencloset-location', format => 'txt' );
        chomp $msg;
        $sent = $self->sms( $phone, $msg );
        $self->log->error("Failed to send SMS: $phone, $msg") unless $sent;

        ## Google Calendar
        my $volunteer = $work->volunteer;
        my $from      = $work->activity_from_date;
        my $to        = $work->activity_to_date;
        my $event_id  = $self->calendar_insert( $volunteer->name, $from, $to );
        $work->update( { event_id => $event_id } );
    }
    elsif ( $status =~ /canceled|drop/ ) {
        my $event_id = $work->event_id;
        $self->delete_event($event_id) if $event_id;

        if ( $from_status eq 'approved' ) {
            my $email = Email::Simple->create(
                header => [
                    From => $self->config->{email_notify_from},
                    To   => $self->config->{email_notify_to},
                    Subject =>
                        sprintf(
                        "[열린옷장 승인된 봉사활동 취소] %s님이 봉사활동을 취소하였습니다.",
                        $volunteer->name ),
                ],
                body => '--',
            );
            $self->send_mail( encode_utf8( $email->as_string ) );
        }
    }

    $self->render( json => { $work->get_columns } );
}

=head2 update_1365

    PUT /works/:id/1365

=cut

sub update_1365 {
    my $self  = shift;
    my $work  = $self->stash('work');
    my $_1365 = $self->param('1365') || 0;

    $work->update( { done_1365 => $_1365 } );
    $self->render( json => { $work->get_columns } );
}

=head2 add_guestbook

    GET /works/:id/guestbook?authcode=xxxx

=cut

sub add_guestbook {
    my $self     = shift;
    my $authcode = $self->param('authcode') || '';
    my $work     = $self->stash('work');

    return $self->error( 400, 'Wrong authcode' ) if $authcode ne $work->authcode;
    $self->render;
}

=head2 create_guestbook

    POST /works/:id/guestbook

=cut

sub create_guestbook {
    my $self     = shift;
    my $authcode = $self->param('authcode') || '';
    my $work     = $self->stash('work');

    return $self->error( 400, 'Wrong authcode' ) if $authcode ne $work->authcode;

    my $name       = $self->param('name');
    my $age_group  = $self->param('age-group');
    my $impression = $self->param('impression');
    my $imprss_etc = $self->param('impression-etc');
    my $activities = $self->every_param('activity');
    my $atvt_etc   = $self->param('activity-etc');
    my $want_to_do = $self->every_param('want-to-do');
    my $todo_etc   = $self->param('want-to-do-etc');
    my $comment    = $self->param('comment');

    my $guestbook = $self->schema->resultset('VolunteerGuestbook')->create(
        {
            volunteer_work_id => $work->id,
            name              => $name,
            age_group         => $age_group,
            impression        => $impression || $imprss_etc,
            activity          => join( '|', @$activities ) || $atvt_etc,
            want_to_do        => join( '|', @$want_to_do ) || $todo_etc,
            comment           => $comment,
        }
    );

    return $self->error( 500, 'Failed to create Volunteer Guestbook' ) unless $guestbook;

    my $email = Email::Simple->create(
        header => [
            From => $self->config->{email_notify_guestbook_from},
            To   => $self->config->{email_notify_guestbook_to},
            Subject =>
                sprintf( "[열린옷장 봉사활동 방명록] %s님의 방명록이 등록되었습니다.", $name ),
        ],
        body => $self->url_for( 'work', { id => $work->id } )->query( authcode => $authcode )->to_abs
    );

    $self->send_mail( encode_utf8( $email->as_string ) );
    $self->render( 'work/thanks', guestbook => $guestbook );
}

=head2 able_hour

    GET /works/hours/:ymd

=cut

sub able_hour {
    my $self = shift;
    my $ymd  = $self->param('ymd');
    $self->render( json => $self->_able_hour($ymd) );
}

sub _validate_volunteer {
    my ( $self, $v ) = @_;

    $v->required('name');
    $v->required('phone')->like(qr/^\d{3}-\d{4}-\d{3,4}$/);
}

sub _validate_volunteer_work {
    my ( $self, $v ) = @_;

    $v->required('activity-datetime')->like(qr/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}-\d{2}:\d{2}$/);
}

sub _able_hour {
    my ( $self, $ymd ) = @_;
    my ( $year, $mm, $dd ) = split /-/, $ymd;

    my $tz     = $self->config->{timezone};
    my $parser = $self->schema->storage->datetime_parser;
    my $dt     = DateTime->new( year => $year, month => $mm, day => $dd, time_zone => $tz );
    my $rs     = $self->schema->resultset('VolunteerWork')->search(
        {
            activity_from_date => {
                -between => [$parser->format_datetime($dt), $parser->format_datetime( $dt->clone->add( days => 1 ) )]
            },
            status => { '!=' => 'canceled' }
        }
    );

    my %schedule;
    while ( my $row = $rs->next ) {
        my $from = $row->activity_from_date;
        my $to   = $row->activity_to_date;
        ## 09 ~ 12 로 신청했을때,
        ## 09:00 ~ 11:59 로 변경해야 시간대별로 계산하기 수월함
        $to->subtract(minutes => 1);
        $schedule{$_}++ for ( $from->hour .. $to->hour );
    }

    my $dow            = $dt->day_of_week;                  # 1-7 (Monday is 1)
    my $max_volunteers = $self->config->{max_volunteers};

    my %result;
    my @templates = qw/09:00-12:00 14:00-17:00 17:00-20:00/;
    for my $template (@templates) {
        my $able = 1;
        my ( $start, $end ) = split /-/, $template;
        $start = substr $start, 0, 2;
        $end   = substr $end,   0, 2;
        $end -= 1;
        for my $hour ( $start .. $end ) {
            my $max
                = defined $max_volunteers->{$dow}{$hour} ? $max_volunteers->{$dow}{$hour} : $max_volunteers->{default};
            if ( $max == 0 || $schedule{$hour} && $schedule{$hour} >= $max ) {
                $able = 0;
                last;
            }
        }

        $result{$template} = $able;
    }

    return {%result};
}

=head2 list

    GET /works

=cut

sub list {
    my $self   = shift;
    my $status = $self->param('status') || 'reported';
    my $query  = $self->param('q') // '';

    $self->stash( pageset => '' );    # prevent undefined error in template

    my $tz = $self->config->{timezone} || 'Asia/Seoul';
    my ( $works, $standby );
    my $parser = $self->schema->storage->datetime_parser;
    $works = $self->schema->resultset('VolunteerWork')
        ->search( { status => $status }, { order_by => 'activity_from_date' } );

    if ( $status eq 'done' ) {
        $works = $works->search(
            {
                activity_from_date =>
                    { '>' => $parser->format_datetime( DateTime->now( time_zone => $tz )->subtract( days => 7 ) ) },
                need_1365 => 1,
                done_1365 => { '<>' => 0 },
            },
            { order_by => [{ -desc => 'need_1365' }, { -asc => 'done_1365' }, { -desc => 'activity_from_date' }] }
        );
        $standby
            = $self->schema->resultset('VolunteerWork')->search( { status => $status, need_1365 => 1, done_1365 => 0, },
            { order_by => [{ -desc => 'need_1365' }, { -asc => 'done_1365' }, { -desc => 'activity_from_date' }] } );
    }
    elsif ( $status eq 'canceled' or $status eq 'drop' ) {
        my $p = $self->param('p') || 1;
        $works = $works->search( undef, { page => $p, rows => 10, order_by => { -desc => 'id' } } );

        my $pageset = Data::Pageset->new(
            {
                total_entries    => $works->pager->total_entries,
                entries_per_page => $works->pager->entries_per_page,
                pages_per_set    => 5,
                current_page     => $p,
            }
        );

        $self->stash( pageset => $pageset );
    }

    if ($query) {
        $works = $self->schema->resultset('VolunteerWork')->search(
            {
                -or => {
                    'volunteer.name'  => $query,
                    'volunteer.phone' => $self->phone_format($query),
                    'volunteer.email' => $query
                }
            },
            { join => 'volunteer' }
        );
    }

    $self->render( works => $works, standby => $standby );
}

1;
